### PR TITLE
[Tradeshield]: Disable close orders, test open orders

### PR DESCRIPTION
**Pull request by amityadav0**
## Description

- [x] Disable close orders for perpetual
- [x] Delete unused code
- [x] Add edge cases such as, user should not have any open position or another order while creating new order
- [x] Test open orders
- [x] Test cancel open orders
- [ ] Test execute open orders

## Testing results

- Open perpetual order
```
ay@Amits-MBP ~ %  elysd tx tradeshield create-perpetual-open-order long 10 2 ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4 1000000ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65 0.5 --from=validator -y --gas=400000
code: 0
codespace: ""
data: ""
events: []
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: '[]'
timestamp: ""
tx: null
txhash: 0B9DA99B00D1D1E056F188F66E23913E14E608E810C2B4E389C65D02652EA713
ay@Amits-MBP ~ % elysd q tradeshield list-pending-perpetual-order                           
pagination:
  next_key: null
  total: "0"
pending_perpetual_order:
- borrow_interest_rate: "0.000000000000000000"
  funding_rate: "0.000000000000000000"
  liquidation_price: "3.936844899406677999"
  perpetual_order:
    collateral:
      amount: "1000000"
      denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    leverage: "10.000000000000000000"
    order_id: "1"
    owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
    perpetual_order_type: LIMITOPEN
    poolId: "2"
    position: LONG
    position_id: "0"
    status: PENDING
    stop_loss_price: "0.000000000000000000"
    take_profit_price: "10000000000000000000000000000000000000000.000000000000000000"
    trading_asset: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
    trigger_price:
      rate: "0.500000000000000000"
      trading_asset_denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
  position_size:
    amount: "2343247"
    denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4

```

- Try to open perpetual order with same pool and address
```
- attributes:
  - index: true
    key: signature
    value: BY/mjkRhProD30ARPq67uD3Cdi6GzBBrAcPBclNVYh5vujajohaEv8uoDqitVKxgmOp0/t4dIm3sCwLbXatCsQ==
  type: tx
gas_used: "38243"
gas_wanted: "400000"
height: "10944882"
info: ""
logs: []
raw_log: 'failed to execute message; message index: 0: user already has a order for
  the same pool: invalid request'
```

- Cancel order
```
ay@Amits-MBP ~ % elysd q tradeshield list-pending-perpetual-order
pagination:
  next_key: null
  total: "0"
pending_perpetual_order:
- borrow_interest_rate: "0.000000000000000000"
  funding_rate: "0.000000000000000000"
  liquidation_price: "3.936844899406677999"
  perpetual_order:
    collateral:
      amount: "1000000"
      denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    leverage: "10.000000000000000000"
    order_id: "1"
    owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
    perpetual_order_type: LIMITOPEN
    poolId: "2"
    position: LONG
    position_id: "0"
    status: PENDING
    stop_loss_price: "0.000000000000000000"
    take_profit_price: "10000000000000000000000000000000000000000.000000000000000000"
    trading_asset: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
    trigger_price:
      rate: "0.500000000000000000"
      trading_asset_denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
  position_size:
    amount: "2343247"
    denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
- borrow_interest_rate: "0.000000000000000000"
  funding_rate: "0.000000000000000000"
  liquidation_price: "0.376692390095044084"
  perpetual_order:
    collateral:
      amount: "1000000"
      denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    leverage: "10.000000000000000000"
    order_id: "2"
    owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
    perpetual_order_type: LIMITOPEN
    poolId: "4"
    position: LONG
    position_id: "0"
    status: PENDING
    stop_loss_price: "0.000000000000000000"
    take_profit_price: "10000000000000000000000000000000000000000.000000000000000000"
    trading_asset: ibc/B4314D0E670CB43C88A5DCA09F76E5E812BD831CC2FEC6E434C9E5A9D1F57953
    trigger_price:
      rate: "0.500000000000000000"
      trading_asset_denom: ibc/B4314D0E670CB43C88A5DCA09F76E5E812BD831CC2FEC6E434C9E5A9D1F57953
  position_size:
    amount: "24489478"
    denom: ibc/B4314D0E670CB43C88A5DCA09F76E5E812BD831CC2FEC6E434C9E5A9D1F57953
ay@Amits-MBP ~ % elysd tx tradeshield cancel-perpetual-order 2 --from=validator --gas=400000 -y
code: 0
codespace: ""
data: ""
events: []
gas_used: "0"
gas_wanted: "0"
height: "0"
info: ""
logs: []
raw_log: '[]'
timestamp: ""
tx: null
txhash: 94BE6E0D27407856EFAAD3BDA2B8C91234A263BC99432F54AC71649DF70D8F10
ay@Amits-MBP ~ % elysd q tradeshield list-pending-perpetual-order                              
pagination:
  next_key: null
  total: "0"
pending_perpetual_order:
- borrow_interest_rate: "0.000000000000000000"
  funding_rate: "0.000000000000000000"
  liquidation_price: "3.936844899406677999"
  perpetual_order:
    collateral:
      amount: "1000000"
      denom: ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65
    leverage: "10.000000000000000000"
    order_id: "1"
    owner_address: elys1g3qnq7apxv964cqj0hza0pnwsw3q920lcc5lyg
    perpetual_order_type: LIMITOPEN
    poolId: "2"
    position: LONG
    position_id: "0"
    status: PENDING
    stop_loss_price: "0.000000000000000000"
    take_profit_price: "10000000000000000000000000000000000000000.000000000000000000"
    trading_asset: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
    trigger_price:
      rate: "0.500000000000000000"
      trading_asset_denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
  position_size:
    amount: "2343247"
    denom: ibc/E2D2F6ADCC68AA3384B2F5DFACCA437923D137C14E86FB8A10207CF3BED0C8D4
```
### What has Changed?

What specific problem were you aiming to address, and how did you successfully resolve it? If tests were not uploaded for this pull request or if coverage decreased, please provide an explanation for the change.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

File: x/tradeshield/migrations/v5_migrator.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ import (
 func (m Migrator) V5Migration(ctx sdk.Context) error {
 
 	m.keeper.DeleteAllPendingSpotOrder(ctx)
+	m.keeper.DeleteAllPendingPerpetualOrder(ctx)
```
Comment (cosmic-vagabond) [inline x/tradeshield/migrations/v5_migrator.go:10]: @amityadav0 you may need to create a new migrator if you want those functions calls to be called during the next testnet upgrade

File: x/tradeshield/types/events.go [hunk 2]
```diff
@@ -7,8 +7,8 @@ import (
 )
 
 const (
-	TypeEvtCloseSpotOrder      = "close_spot_order"
-	TypeEvtClosePerpetualOrder = "close_perpetual_order"
+	TypeEvtCloseSpotOrder       = "close_spot_order"
+	TypeEvtCancelPerpetualOrder = "cancel_perpetual_order"
```
Comment (cosmic-vagabond) [inline x/tradeshield/types/events.go:11]: @amityadav0 can you clarify why we renaming the close perpetual order to cancel perpetual order?

File: x/tradeshield/migrations/v5_migrator.go (hunk 1 above)
Comment (amityadav0) [reply x/tradeshield/migrations/v5_migrator.go:10]: I created them in this PR https://github.com/elys-network/elys/pull/912/files, this is not yet included, is that right ?

File: x/tradeshield/types/events.go (hunk 2 above)
Comment (amityadav0) [reply x/tradeshield/types/events.go:11]: to avoid confusion, so we can close perpetual order which is already open

but this event is used  in cancelling a perpetual open order request. It is not used in closing a perpetual position function

does this make sense ? Can change it something else too ?

File: x/tradeshield/migrations/v5_migrator.go (hunk 1 above)
Comment (cosmic-vagabond) [reply x/tradeshield/migrations/v5_migrator.go:10]: you are right, we are good then ;-)

File: x/tradeshield/types/events.go (hunk 2 above)
Comment (cosmic-vagabond) [reply x/tradeshield/types/events.go:11]: they are two separate flows:

1) cancel order: to remove an existing perpertual order from the order book (open or close order), the user decides to cancel a pending order basically

2) close order: to execute the closure of an open perpetual position at a given price point or at current market price (limit close or market close)

given this knowledge, can you clarify which one are you trying to implement here?

Comment (amityadav0) [reply x/tradeshield/types/events.go:11]: 1. cancel order, the above event was use in cancel order as a close order event

Comment (amityadav0) [reply x/tradeshield/types/events.go:11]: https://github.com/elys-network/elys/blob/a644435ff2a6155c2b084114c4113bba6e0561a2/x/tradeshield/keeper/msg_server_perpetual_order.go#L126

**Review by cosmic-vagabond — APPROVED** [review]
